tests/tensorbox/test_tensorbox.py

```diff
@@ -0,0 +1,541 @@
+
+    def test_transpose(self):
+        """Test that the transpose is correct"""
+        x = np.array([[1, 2], [3, 4]])
+        xT = qml.TensorBox(x)
+
+        assert np.all(xT.T == x.T)
+
+
+class TestAutogradBox:
+    """Tests for the AutogradBox class"""
+
+    @pytest.fixture
+    def autograd(self):
+        autograd = pytest.importorskip("autograd")
+        return autograd
+
+    def test_creation(self, autograd):
+        """Test that a AutogradBox is automatically created from a PennyLane numpy tensor"""
+        from pennylane.tensorbox.autograd_box import AutogradBox
+
+        x = qml.numpy.array([0.1, 0.2, 0.3])
+        res = qml.TensorBox(x)
+        assert isinstance(res, AutogradBox)
+        assert res.interface == "autograd"
+        assert isinstance(res.unbox(), np.ndarray)
+        assert np.all(res.unbox() == x)
+
+    def test_len(self, autograd):
+        """Test length"""
+        np = qml.numpy
+        x = np.array([[1, 2], [3, 4]])
+        res = qml.TensorBox(x)
+        assert len(res) == len(x) == 2
+
+    def test_ufunc_compatibility(self, autograd):
+        """Test that the AutogradBox class has ufunc compatibility"""
+        np = qml.numpy
+        x = np.array([0.1, 0.2, 0.3])
+        res = np.sum(np.sin(qml.TensorBox(x)))
+        assert res.unbox().item() == np.sin(0.1) + np.sin(0.2) + np.sin(0.3)
+
+    def test_multiplication(self, autograd):
+        """Test multiplication between tensors and arrays"""
+        np = qml.numpy
+        x = np.array([[1, 2], [3, 4]])
+        y = np.array([[1, 0], [0, 1]])
+
+        xT = qml.TensorBox(x)
+        res = xT * y
+        assert np.all(res.unbox() == x * y)
+
+        yT = qml.TensorBox(y)
+        res = x * yT
+        assert np.all(res.unbox() == x * y)
+
+        res = xT * yT
+        assert np.all(res.unbox() == x * y)
+
+    def test_unbox_list(self, autograd):
+        """Test unboxing a mixed list works correctly"""
+        np = qml.numpy
+        x = np.array([[1, 2], [3, 4]])
+        y = np.array([[1, 0], [0, 1]])
+
+        xT = qml.TensorBox(x)
+        res = xT.unbox_list([y, xT, x])
+
+        assert np.all(res == [y, x, x])
+
+    def test_numpy(self, autograd):
+        """Test that calling numpy() returns a NumPy array representation
+        of the TensorBox"""
+        x = qml.numpy.array([[1, 2], [3, 4]])
+        xT = qml.TensorBox(x)
+        assert isinstance(xT.numpy(), np.ndarray)
+        assert not isinstance(xT.numpy(), qml.numpy.tensor)
+        assert np.all(xT.numpy() == x)
+
+    def test_shape(self, autograd):
+        """Test that arrays return the right shape"""
+        np = qml.numpy
+        x = np.array([[[1, 2], [3, 4]]])
+        x = qml.TensorBox(x)
+        res = x.shape
+        assert res == (1, 2, 2)
+
+    def test_expand_dims(self, autograd):
+        """Test that dimension expansion works"""
+        from pennylane.tensorbox.autograd_box import AutogradBox

```

Comment:
Could we just import these at the class level, rather than in every test?

Reply:
Unfortunately not; python executes all class attributes at module import, which would then fail if the corresponding package was not installed

Reply:
I've split this into separate files, one for each class. This way we can do the imports at the top.